### DISCUSSION TITLE
How to filter traces by request header

### DISCUSSION THREAD

**Discussion by raagulrakesh** [grafana/tempo#3551] [Q&A]
Hi ,
I have enabled traces in tempo using opentelemetery. I want to filter traces by request header but header not showing in traces filter. I have passing request header by `OTEL_INSTRUMENTATION_HTTP_CAPTURE_HEADERS_SERVER_REQUEST`. 

tempo configuration

{
	"overrides.yaml": "overrides:
		  defaults:
		     global:
		       max_bytes_per_trace: 10000000
		",
	"tempo.yaml": "multitenancy_enabled: false
		usage_report:
		  reporting_enabled: true
		compactor:
		  compaction:
		    block_retention: 720h
		distributor:
		  receivers:
		        jaeger:
		          protocols:
		            grpc:
		              endpoint: 0.0.0.0:14250
		            thrift_binary:
		              endpoint: 0.0.0.0:6832
		            thrift_compact:
		              endpoint: 0.0.0.0:6831
		            thrift_http:
		              endpoint: 0.0.0.0:14268
		        opencensus: null
		        otlp:
		          protocols:
		            grpc:
		              endpoint: 0.0.0.0:4317
		            http:
		              endpoint: 0.0.0.0:4318
		ingester:
		      {}
		server:
		      http_listen_port: 3100
		storage:
		      trace:
		        backend: local
		        local:
		          path: /var/tempo/traces
		        wal:
		          path: /var/tempo/wal
		querier:
		      {}
		query_frontend:
		      {}
		overrides:
		      per_tenant_override_config: /conf/overrides.yaml
		"
}

grafana tempo search

![Screenshot 2024-04-09 114231](https://github.com/grafana/tempo/assets/44870259/795acb60-ba7b-48ac-ab7f-8cb70dd761bf)

**Comment by joe-elliott** [comment]
If the header is an [array type](https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/common/v1/common.proto#L36) we currently do not have the ability to query it. The team is working on vparquet4 actively which will extend the parquet format to support queries on these data types.

The syntax for querying arrays is defined here:

https://github.com/grafana/tempo/blob/main/docs/design-proposals/2023-11%20TraceQL%20Extensions.md#arrays

After vp4 is merged we will be able to add these extensions.

**Comment by raagulrakesh** [comment]
I have tried querying array still I didn't get result for the query

![Screenshot 2024-04-11 112109](https://github.com/grafana/tempo/assets/44870259/abcfa831-aeba-4f56-9c0c-b6a5ddee95c9)
![Screenshot 2024-04-11 112136](https://github.com/grafana/tempo/assets/44870259/b7aab9a6-0283-4c45-9de3-4a0a0ace9834)

**Comment by joe-elliott** [comment]
Arrays are not currently supported. The doc i linked to you shows our proposed syntax. We are actively working on merging vp4 and adding support for arrays.

cc @stoewer

**Comment by raagulrakesh** [comment]
Hi,
When can we expect this in live?

Thanks in advance.

**Reply by joe-elliott** [reply]
Hoping to tackle it in the next few months.

**Reply by raagulrakesh** [reply]
Is there any other options to filter temporally for now

**Reply by joe-elliott** [reply]
Currently you would need a different attribute to query. You could likely use the otel transform processor in either [Grafana Alloy](https://github.com/grafana/alloy) or the OTel Collector to write one automatically in your pipeline:

https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/transformprocessor

**Comment by guptamohit24** [comment]
Hello @joe-elliott 

I'm also facing this issue.

To easily search a specific trace of a http request, I exposed a custom header in span using open telemetry HTTP configuration.
In Java spring, header is a multi value map, hence the data type is Array. 

Do you have any update when filtering on Array type will be available ?

**Reply by joe-elliott** [reply]
querying on arrays is available now. if `http.request.header.foo` is an array and it contains the value "bar" then the following query will return it.

```
{ span.http.request.header.foo = "bar" }
```

we have considered more advanced functionality but we feel this covers nearly all use cases and we didn't want to commit to a more complex syntax until we had a better understanding of the needs this does not cover.

**Reply by guptamohit24** [reply]
Hi @joe-elliott from which version of grafana tempo this feature is available?

**Reply by joe-elliott** [reply]
you need vParquet4 blocks which were added in 2.5 and made default in 2.6
